### ISSUE TITLE
Inconsistent behavior  python 2.7.9 vs 2.7.6

### ISSUE THREAD

**Issue by bkuberek** [chimpler/pyhocon#29]
pyhocon==0.3.3

I've noticed that the following is happening for me when deploying my app to an ubuntu server. On my mac running python 2.7.9 the output is fine, but when running in my ubuntu server on python 2.7.6 I get wrong output. I'm still trying to debug this.

Configuration
```
# config/people.conf
{
  database {
    host = "localhost"
    port = "5432"
    user = "people"
    name = "peopledb"
    pass = "dev"
    url = "postgresql://"${database.user}":"${database.pass}"@"${database.host}":"${database.port}"/"${database.name}

    host = ${?DB_HOST}
  }

  include file('user.conf')
}
```
```
# config/user.conf
{
  database {
    host = "192.168.33.10"
  }
}
```

Python 2.7.9 (OS X and homebrew)
```
pyhocon -i config/people.conf
{
  "database": {
    "host": "192.168.33.10",
    "port": "5432",
    "user": "people",
    "name": "peopledb",
    "pass": "dev",
    "url": "postgresql://people:dev@192.168.33.10:5432/peopledb"
  }
}
```

Python 2.7.6 (Ubuntu trusty)
```
pyhocon -i config/people.conf
{
  "database": {
    "host": "192.168.33.10",
    "port": "5432",
    "user": "people",
    "name": "peopledb",
    "pass": "dev",
    "url": "postgresql://people:dev@localhost:5432/peopledb"
  }
}
```

Notice how the `hostname` part of the `url` differs.

**Comment by bkuberek** [comment]
nevermind. I have weird issues with my environment. this is working as expected.

**Comment by bkuberek** [comment]
This is not actually true. I have just compared the output runing 2.7.9 and 2.7.6 on OS X and output is fine.

What is the best way for me to debug this?

**Comment by darthbear** [comment]
I'm not sure. If you put the include first, do you still have the issue? I'll take a look at it tonight.

**Comment by bkuberek** [comment]
If i put the include first wouldn't it be overriden? I am adding some logging around ConfigParser to see if I can figure something out.

**Comment by bkuberek** [comment]
So I figured out I had an environment variable set that was writing to the `DB_HOST` variable, so at least now they are both consistent.
 
The issue here is that the variable `database.url` is consuming the other variables before the rest of the configuration is loaded, thus, getting the value `localhost` as opposed to the value defined in the last include file.

In other words, `database.url` consumes `database.host` and the other variables, before the final override happens. The variable `database.host` is eventually updated to the final value but `database.url` isn't reflecting that.

Should I change the title and description of this issue?

**Comment by darthbear** [comment]
oh ok. Sorry for my previous comment. I understand the problem now. Yes please. I'll try to test it on trusty. So just to make sure, you should see the problem if you do:
```
a {
  b = 5
  c = ${a.b}
}

a {
  b = 7
}
``` 

I'll take a look tonight. Thanks!

**Comment by bkuberek** [comment]
I was about to show you an example but I am once again struggling to reproduce.

I tried these below and they are all working as expected.
```
# db1.conf
{
  host = "localhost"
  port = 8000
  url = ${host}":"${port}
  
  host = "otherhost"
  port = 433
}
```

```
# db2.conf
{
  database {
    host = "localhost"
    port = 8000
    url = ${database.host}":"${database.port}
  }

  database {
    host = "otherhost"
    port = 433
  }
}
```

```
# db3.conf
{
  database {
    host = "localhost"
    port = 8000
    url = ${database.host}":"${database.port}
  }

  include file("db4.conf")
}
```
```
# db4.conf
{
  database {
    host = "otherhost"
    port = 433
  }
}
```

The result is 
```
{
  "database": {
    "host": "other.host.net",
    "port": 433,
    "url": "otherhost:433"
  }
}
```

**Comment by darthbear** [comment]
ok. Thanks for the new results.
So from your first comment, it looks like `host = ${?DB_HOST}` might be the culprit

**Comment by bkuberek** [comment]
here is something:

config:
```
{
  database {
    host = "localhost"
    port = 8000
    url = ${database.host}":"${database.port}
  }

  database {
    host = ${?DB_HOST}
  }

  database {
    host = "other.host.net"
    port = 433
  }
}
```
result
```
{
  "database": {
    "host": "localhost",
    "port": 433,
    "url": "other.host.net:433"
  }
}
```

**Comment by darthbear** [comment]
Thanks for the example. That will definitely help!

**Comment by bkuberek** [comment]
Here is a different configuration, similar error but the error is on the `url` instead:

```
# database.conf
{
  include file("default.conf")
  include file("user.conf")
}
```
```
# default.conf
{
  database {
    host = "localhost"
    port = 8000
    url = ${database.host}":"${database.port}

    host = ${?DB_HOST}
  }
}
```
```
# user.conf
{
  database {
    host = "other.host.net"
    port = 433
  }
}
```

The result is :

```
pyhocon -i database.conf
{
  "database": {
    "host": "other.host.net",
    "port": 433,
    "url": "localhost:8000"
  }
}
```

**Comment by darthbear** [comment]
I think I understand what happened. So when we encounter a substitution, we keep the substitution in a list (substitutions) and a reference to it (so the substitution `${?DB_HOST}` will keep a reference to `database['host']`). At the end we resolve the references and we replace the references with the resolved substitution. But in our case, we create a substitution with `database.host` and then we override it but we kept it in the substitution list. And then when we resolved it, we replaced `database.host`.
The way to fix it would be to instead of considering the substitutions list to consider a dictionary path -> substitution instead so in our case we can see that the substitution is overriden and remove it.

**Comment by bkuberek** [comment]
This sounds like might take a bit of work to fix. I don't think I much help in the implementation details yet. But I am able to work with this as long as I have the right order in my configuration file. That is ok for now so this is not blocking me anymore. I appreciate the python port and your effort driving this.

**Comment by darthbear** [comment]
Thank you for your help. Yes that will take some time to fix and test it. I'll work on it over the weekend and let you know. Thanks for your help pinpointing the problem!

**Comment by bkuberek** [comment]
thanks!

**Comment by darthbear** [comment]
The bug should be fixed now in `pyhocon==0.3.4`. Thanks @bkuberek for your help!
